runtime/runtime_events.c

```diff
@@ -404,20 +396,37 @@ static void runtime_events_create_raw(void) {
   }
 }
 
+/* create/teardown STWs
+
+   The STW API does have an enter barrier before the handler code is
+   run, however, the enter barrier itself calls the runtime events API
+   after arrival. Thus, the barrier in the STWs below is needed both
+   to ensure that all domains have actually reached the handler before
+   we start/stop (and are not calling the runtime events API from the
+   STW code), and of course to ensure that the setup/teardown is
+   observed by all domains returning from the STW. */
+
```

Comment:
I think it’s great to have such explanation comments, but, maybe because I’m not familiar with runtime events, I’m having a hard time understanding this one. By “runtime events API”, do we mean e.g. `CAML_EV_BEGIN`? If so, this function is called before arriving to the barrier, e.g. at https://github.com/ocaml/ocaml/pull/12579/commits/8d2ffa556897e2a5ed2f95b0e34ec9e193e71399#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR1400, and I don’t understand how the barrier in the function below helps with that.

If I’m misunderstanding, please, let me know.

Reply:
The point is to avoid a race there. Calling the STW API before the enter barrier is fine because it's either properly initialised or not, but calling it after can race with the actual setup. I'll try and clarify.

Reply:
I understand better now and find the comment clearer, thanks.